publish/action.yml

```diff
@@ -0,0 +1,55 @@
+name: 'Publish to PyPI'
+description: 'Publishes an artifact to PyPI'
+inputs:
+  user:
+    description: 'The username to use to publish to PyPI. Defaults to `__token__`.'
+    required: false
+    default: '__token__'
+  password:
+    description: 'The password to use to publish to PyPI'
+    required: true
+  repository_url:
+    description: 'The PyPI repository URL to publish to'
+    required: false
+  test_user:
+    description: 'The username to use when test is set to true. Uses the value of `user` if not provided.'
+    required: false
+  test_password:
+    description: 'The password to use when test is set to true. Uses the value of `password` if not provided.'
+    required: false
+  test_repository_url:
+    description: 'The PyPI repository URL to publish to when test is set to true. Defaults to `https://test.pypi.org/legacy/`'
+    required: false
+    default: 'https://test.pypi.org/legacy/'
+  test:
+    description: 'Whether to publish to the test repository or the production repository'
+    required: false
+    default: false
+  verify_metadata:
+    description: 'Whether to check package metadata before uploading'
+    required: false
+    default: true
+  skip_existing:
+    description: 'Do not fail if a Python package distribution exists in the target package index'
+    required: false
+    default: false
+  verbose:
+    description: 'Show verbose output.'
+    required: false
+    default: false
+runs:
+  using: 'composite'
+  steps:
+    - uses: actions/download-artifact@v2
+      with:
+        name: artifact
+        path: dist
+
+    - uses: pypa/gh-action-pypi-publish@v1.4.2
+      with:
+        user: ${{ (inputs.test && (inputs.test_user || inputs.user)) || inputs.user }}
+        password: ${{ (inputs.test && (inputs.test_password || inputs.password)) || inputs.password }}

```

Comment:
I don't quite grok this. Is this allowing maintainer to pass in password as plain text?

Reply:
Currently, by design, the only way to pass inputs to a step is in plain text; a custom step cannot access secrets from the repository that is executing it.

This logic allows the user to specify a different username & password to use when the `test` property is set to `true` (since this will publish to the specified `test_repository_url`, which presumably will require different credentials than the production `repository_url`). So for example, if `test` is set to `true` and `test_password` was provided, it will be used instead of `password`, but if `test` is set to `true` and `test_password` was _not_ provided, then `password` will be used.

Reply:
Isn't passing password in plain text a security risk?

Reply:
What attack scenario are you envisioning? In general, yes, but I'm not sure this advice applies in the context of custom Github actions, especially considering that this is how every custom Github action passes passwords, since there is no other way to do so.

Reply:
I see what you mean, but staff should be using [our workflow template](https://github.com/zanecodes/.github/blob/refactor/workflow-templates/publish-to-pypi.yml), which takes the password from their repository secrets (or the `spacetelescope` organization secrets) and passes it into the step, they should almost never need to modify the workflow themselves. I agree that this is somewhat of a design flaw in custom Github Actions though, although it [doesn't seem like Github intends to change this behavior](https://github.community/t/accessing-secrets-of-custom-actions-repo/126280) anytime soon.

Reply:
We should not give staff a gun to shoot themselves in the foot, but I have given my review and it is up to you to accept the comment or not. So, I am going to shuddup now. 😸

Reply:
It's worth noting that the currently published version of `spacetelescope/action-publish_to_pypi` [already works this way](https://github.com/spacetelescope/.github/blob/master/workflow-templates/publish-to-pypi.yml#L11-L14) anyhow, so staff have had this particular footgun available for quite some time.

Reply:
😬

Reply:
I could envision an argument that allowing custom actions to access secrets from the repository that's using them would itself be a security risk, as users of the custom step wouldn't have an easy way to tell what secrets the custom step is reading without referencing its source code, and malicious maintainers could update the custom step to exfiltrate any and all secrets from the calling repositories at any time. By forcing users to explicitly pass in secrets, it at least ensures that they are aware of which ones are accessible by the custom step.

That said, it would be much nicer if custom steps allowed some syntax akin to this, thus forcing users to put secret values into actual Github secrets if they wish to pass them to the custom step, while still maintaining explicit access to only a restricted set of secrets specified by the user:
```
- name: custom action
  uses: spacetelescope/action-publish_to_pypi@master
  secrets:
    user: PYPI_USERNAME_STSCI_MAINTAINER
    password: PYPI_PASSWORD_STSCI_MAINTAINER
```

Reply:
Hmm... Would it be useful to contact GitHub Help Desk to ask about this and see what they say?

Reply:
Perhaps, but in all honesty I doubt they'll give us any new information that their [official documentation doesn't already say](https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow)

Reply:
Apologies, I was logged in using the wrong account so I'm reposting my first comment:

Currently, by design, the only way to pass inputs to a step is in plain text; a custom step cannot access secrets from the repository that is executing it.

This logic allows the user to specify a different username & password to use when the `test` property is set to `true` (since this will publish to the specified `test_repository_url`, which presumably will require different credentials than the production `repository_url`). So for example, if `test` is set to `true` and `test_password` was provided, it will be used instead of `password`, but if `test` is set to `true` and `test_password` was _not_ provided, then `password` will be used.

Reply:
For now, I'll add a comment to the workflow template warning developers not to hardcode secrets into the workflow file, and perhaps a `TODO` for once Github provides an alternative.

Reply:
Sounds good to me. Thanks!